Avoid dead-lock in `util/sliding_window`

vickenty opened the pull request:
### What does this PR do?

Fix a dead-lock in the sliding window implementation during agent shutdown.

### Motivation

Make agent stop more reliably.

### Additional Notes

Sliding window periodically calls user function `statsUpdateFn` while holding a read-only lock (Lock A).

`statsUpdateFn` then retrieves the accumulated average by calling Average (RLock B).

If between the two calls above a call to `Stop` is made, a pending write lock would block RLock B, while RLock A would prevent the write lock inside `Stop`, leading to a deadlock.

This PR avoids the issue by computing the average upfront and passing it to the statsUpdateFn.

### Describe how to test/QA your changes

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

vickenty committed:
Avoid dead-lock in sliding_window

Sliding window periodically calls user function statsUpdateFn while
holding a read-only lock (Lock A).

statsUpdateFn then retrieves the accumulated average by calling
Average (RLock B).

If between the two calls above a call to Stop is made, a pending write
lock would block RLock B, while RLock A would prevent write Lock
inside Stop, leading to a deadlock.

Avoid the issue by computing average upfront and passing it to the
statsUpdateFn - it has only one method to call anyway.

sgnn7 commented on pkg/collector/worker/utilization_tracker.go:
```diff
@@ -117,9 +117,7 @@ func newUtilizationTrackerWithClock(
-	ut.statsUpdateFunc = func(sw util.SlidingWindow) {
-		utilization := sw.Average()
-
+	ut.statsUpdateFunc = func(utilization float64) {
```
Not a blocker and would be a bigger change but I wonder if `statsUpdateFunc` is a good name anymore that we always only return the average.

sgnn7 reviewed (commented):
@vickenty Left a couple of comments

vickenty committed:
Cleanup redundant lock

Bucket update and average calculation now only ever happen
sequentially on the same ticker goroutine, so we don't need a lock to
synchronize their access to buckets.

vickenty commented on pkg/util/sliding_window.go:
```diff
@@ -214,19 +210,11 @@ func (sw *slidingWindow) WindowSize() time.Duration {
+func (sw *slidingWindow) average() float64 {
 	totalVal := 0.0
 
-	sw.bucketsLock.RLock()
```
Bucket update and average calculation now happen sequentially on the same goroutine, so they cannot race. In fact, I think the entire lock is redundant now, thus 96bf875.
